includes/class-sensei-guest-user.php

```diff
@@ -0,0 +1,89 @@
+<?php
+/**
+ * Guest User
+ *
+ * Handles operations related to allowing guest users take a course.
+ *
+ * @package Sensei\Frontend
+ * @since 1.3.0
+ */
+
+if ( ! defined( 'ABSPATH' ) ) {
+	exit; // Exit if accessed directly.
+}
+
+/**
+ * Sensei Guest User Class.
+ *
+ * @author Automattic
+ *
+ * @since $$next-version$$
+ * @package Core
+ */
+class Sensei_Guest_User {
+
+	/**
+	 * Keeps a reference to the Guest Student role object
+	 *
+	 * @access protected
+	 * @since $$next-version$$
+	 *
+	 * @var string
+	 */
+	protected $guest_student_role = 'guest_student';
+
+	/**
+	 * Sensei_Guest_User constructor.
+	 *
+	 * @since $$next-version$$
+	 */
+	public function __construct() {
+
+		// Check if the Guest Student role exists.
+		$guest_role = get_role( $this->guest_student_role );
+
+		// If Guest Student is not a valid WordPress role create it.
+		if ( ! is_a( $guest_role, 'WP_Role' ) ) {
+			// Create the role.
+			add_role( $this->guest_student_role, __( 'Guest Student', 'sensei-lms' ) );
+		}
+
+		add_action( 'wp', array( $this, 'sensei_create_guest_user_for_open_course' ), 9 );
+	}
+
+	/**
+	 * Create a guest user for open access courses if no user is logged in.
+	 *
+	 * @since  $$next-version$$
+	 */
+	public function sensei_create_guest_user_for_open_course() {
+		global $post;
+
+		// Conditionally create Guest Student user and set role for open course.
+		if (
+			is_singular( 'course' )
+			&& isset( $_POST['course_start'] )
+			&& isset( $_POST['woothemes_sensei_start_course_noonce'] )
+			&& wp_verify_nonce( wp_unslash( $_POST['woothemes_sensei_start_course_noonce'] ), 'woothemes_sensei_start_course_noonce' ) // phpcs:ignore WordPress.Security.ValidatedSanitizedInput.InputNotSanitized -- Don't modify the nonce.
+			&& ! post_password_required( $post->ID )
+			&& ! is_user_logged_in()
+			&& get_post_meta( $post->ID, 'open_access', true )
+		) {
```

Comment:
Both for clean code reasons and because we'll probably be extending the conditions here to include some lesson actions, WDYT of breaking out some parts to methods or at least variables — we could have a higher-level layer of what's happening:

```php
if ( $is_open_access_course && $this->is_take_course_action() && .. ) { 
  $user_id = $this->create_guest_user();
  $this->login_user( $user_id );
}
```

Reply:
You are totally right. I've broken the operations down in multiple methods [here](https://github.com/Automattic/sensei/pull/6259/commits/d3a1dbd1b51fa2ab1eca9b2ee64a32ef9b1b6fd6). Also, I moved the open access check to a separate method because I didn't want to run an extra meta query in case the if clause returns false earlier. Thanks!